src/protocol/boolean/generate_random_bits/malicious.rs

```diff
@@ -20,7 +20,7 @@ impl<F: Field> RandomBits<F> for MaliciousContext<'_, F> {
         let ctx = &c;
         let malicious_triples =
             try_join_all(triples.into_iter().enumerate().map(|(i, t)| async move {
-                ctx.upgrade_bit_triple(&BitOpStep::from(i), record_id, t)
+                ctx.upgrade_for_record_with(&BitOpStep::from(i), record_id, t)
```

Comment:
This one seems different from all the others. Is this the only place where the client needs to provide a record_id when upgrading?

Reply:
Yes... this is the only place where the entire input isn't presented for upgrade all at once (it's not really an input at all -- it's randomness I think used for bit decomposition), so it's not possible to refactor it such that the record IDs are generated internal to the upgrade implementation for the thing being upgraded.

It might be possible to refactor this to `upgrade_for_record<T>(record_id, input: T)` with  `T = Vector<BitConversionTriple>` (getting rid of the `_with`) if that's preferable.

Reply:
Not really worth holding this change up.  Want to open an issue so that we can capture any ideas you have and maybe revisit later?